libbpf-tools/cpufreq.c

```diff
@@ -0,0 +1,196 @@
+"\n";
+
+static const struct argp_option opts[] = {
+	{ "duration", 'd', "DURATION", 0, "Duration to trace"},
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'd':
+		errno = 0;
+		env.duration = strtol(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "Invalid duration: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int nr_cpus;
+
+static int open_and_attach_perf_event(int freq, struct bpf_program *prog,
+				struct bpf_link *links[])
+{
+	struct perf_event_attr attr = {
+		.type = PERF_TYPE_SOFTWARE,
+		.freq = 1,
+		.sample_period = freq,
+		.config = PERF_COUNT_SW_CPU_CLOCK,
+	};
+	int i, fd;
+
+	for (i = 0; i < nr_cpus; i++) {
+		fd = syscall(__NR_perf_event_open, &attr, -1, i, -1, 0);
+		if (fd < 0) {
+			fprintf(stderr, "failed to init perf sampling: %s\n",
+				strerror(errno));
+			return -1;
+		}
+		links[i] = bpf_program__attach_perf_event(prog, fd);
+		if (libbpf_get_error(links[i])) {
+			fprintf(stderr, "failed to attach perf event on cpu: "
+				"%d\n", i);
+			links[i] = NULL;
+			close(fd);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+static void print_linear_hists(struct bpf_map *hists,
+			struct cpufreq_bpf__bss *bss)
+{
+	struct hkey lookup_key = {}, next_key;
+	int err, fd = bpf_map__fd(hists);
+	struct hist hist;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return;
+		}
+		print_linear_hist(hist.slots, MAX_SLOTS, 0, 200, next_key.comm);
+		printf("\n");
+		lookup_key = next_key;
+	}
+
+	printf("\n");
+	print_linear_hist(bss->syswide.slots, MAX_SLOTS, 0, 200, "syswide");
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct bpf_link **links = NULL;
+	struct cpufreq_bpf *obj;
+	int err, i;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = cpufreq_bpf__open_and_load();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF object\n");
+		return 1;
+	}
+
+	nr_cpus = libbpf_num_possible_cpus();
+	if (nr_cpus > MAX_CPU_NR) {
+		fprintf(stderr, "The number of cpu cores is too much, please "
+			"increase MAX_CPU_NR's value and recompile");
+		goto cleanup;
```

Comment:
if nr_cpus is bigger than MAX_CPU_NR, in cleanup you'll go out of bounds of links array

Reply:
and links is not even allocated at that point yet

Reply:
Oh, fixed thanks! I realized that there was a similar problem with another tool submitted before, I will submit another PR to fix that tool.